File: javascript/negamax.js
/*
  Perfect Tic-Tac-Toe player in JavaScript
  Copyright © 2012 Bart Massey
  [This program is licensed under the "MIT License"]
  Please see the file COPYING in the source
  distribution of this software for license terms.
*/

/*
  Pseudocode:

  To calculate the negamax value of a position s:
     If s is a final state (game over)
         return the value of s to the side on move
     v <- -1
     for each legal move m in s
        s' <- m(s)
        v' <- -negamax(s')
        if v' > v
            v <- v'
     return v
*/

function negamax(onmove, board) {
    var v = gamevalue(onmove, board)
    if (v != -2)
        /* game is over */
        return v
    v = -1
    for (var r = 0; r < 3; r++) {
        for (var c = 0; c < 3; c++) {
            if (board[r][c] == 0) {
                board[r][c] = onmove
                var v0 = -negamax(-onmove, board)
                if (v0 > v)
                    v = v0
                board[r][c] = 0
            }
        }
    }
    return v
}

File: haskell/Negamax.hs
-- Copyright © 2015 Bart Massey
-- [This program is licensed under the "MIT License"]
-- Please see the file COPYING in the source
-- distribution of this software for license terms.

-- Perfect Tic-Tac-Toe negamax in Haskell

{-
  Pseudocode:

  To calculate the negamax value of a position s:
     If s is a final state (game over)
         return the value of s to the side on move
     v <- -1
     for each legal move m in s
        s' <- m(s)
        v' <- -negamax(s')
        if v' > v
            v <- v'
     return v
-}

module Negamax
where

import Control.Monad
import Data.Array.IO
import Data.IORef

import GameValue

negamax :: Int -> IOUArray (Int, Int) Int -> IO Int
negamax onMove board = do
  v0 <- gameValue onMove board
  if v0 /= -2
  then
      -- game is over
      return v0
  else do
    v <- newIORef (-1)
    forM_ [0..2] $ \r -> do
      forM_ [0..2] $ \c -> do
        cell <- readArray board (r, c)
        when (cell == 0) $ do
          writeArray board (r, c) onMove
          vn <- negamax (-onMove) board
          v' <- readIORef v
          when ((-vn) > v') $
              writeIORef v (-vn)
          writeArray board (r, c) 0
    readIORef v


testBoard :: IO (IOUArray (Int, Int) Int)
testBoard = newListArray ((0,0),(2,2)) [1,1,-1,-1,-1,1,1,1,0]
